File: apps/web/src/core/domain/ticket.ts
import type { Branded } from './branded'
import type { ProjectId, UserId } from './project'
import type { Priority } from './todo'

export type TicketId = Branded<string, 'TicketId'>

export const ticketId = (value: string): TicketId => value as TicketId

export const TICKET_STATUSES = ['open', 'in_progress', 'blocked', 'closed', 'cancelled'] as const

export type TicketStatus = (typeof TICKET_STATUSES)[number]

export const WAYFINDER_TYPES = ['map', 'research', 'prototype', 'grilling', 'task'] as const

export type WayfinderType = (typeof WAYFINDER_TYPES)[number]

export type Ticket = {
	readonly id: TicketId
	readonly projectId: ProjectId
	readonly parentId: TicketId | undefined
	readonly slug: string
	readonly title: string
	readonly body: string
	readonly status: TicketStatus
	readonly priority: Priority
	readonly assignee: UserId | undefined
	readonly tags: readonly string[]
	readonly externalRef: string
	readonly dependsOn: readonly TicketId[]
	readonly wayfinder: WayfinderType | undefined
	readonly createdBy: UserId | undefined
	readonly createdAt: Date
	readonly updatedAt: Date
}

export const isTerminal = (status: TicketStatus): boolean => status === 'closed' || status === 'cancelled'

File: packages/folio-core/domain/ticket.go
package domain

import "time"

type TicketStatus string

const (
	TicketOpen       TicketStatus = "open"
	TicketInProgress TicketStatus = "in_progress"
	TicketBlocked    TicketStatus = "blocked"
	TicketClosed     TicketStatus = "closed"
	TicketCancelled  TicketStatus = "cancelled"
)

func (s TicketStatus) IsTerminal() bool {
	return s == TicketClosed || s == TicketCancelled
}

type WayfinderType string

const (
	WayfinderMap       WayfinderType = "map"
	WayfinderResearch  WayfinderType = "research"
	WayfinderPrototype WayfinderType = "prototype"
	WayfinderGrilling  WayfinderType = "grilling"
	WayfinderTask      WayfinderType = "task"
)

type Ticket struct {
	ID          TicketID
	ProjectID   ProjectID
	ParentID    TicketID
	Slug        string
	Title       string
	Body        string
	Status      TicketStatus
	Priority    Priority
	Assignee    UserID
	Tags        []string
	ExternalRef string
	DependsOn   []TicketID
	Wayfinder   WayfinderType
	CreatedBy   UserID
	CreatedAt   time.Time
	UpdatedAt   time.Time

	Progress Progress
	Blocked  bool
	Cycle    int
	Phase    Phase
}

type TicketFilter struct {
	ParentID TicketID
	Status   []TicketStatus
	Priority Priority
	Assignee UserID
	Tags     []string
	Search   string
	Limit    int
	Offset   int
}

type TicketBrief struct {
	Ticket  Ticket
	Plans   []Plan
	Todos   []Todo
	Journal []JournalEntry
	Docs    []Doc
	Cycles  []Cycle
}
